### ISSUE TITLE
Two bugs with tactics

### ISSUE THREAD

**Issue by parno** [FStarLang/FStar#1415]
Hi all,

We've run into two tactic-related issues with Vale.  In both cases, we're trying to use the int canonicalizer.  You can reproduce them by checking out the `fast_mul` branch of the Vale repo (https://github.com/project-everest/vale) at commit `b477a46`.  In the base of the repo, run scons, which will take a few minutes (note that if you haven't built Vale before, you'll first need to run `nuget.exe restore ./tools/Vale/src/packages.config -PackagesDirectory tools/FsLexYacc`).  

With the version that's checked in, you should see a tactic failure:
`obj/crypto/aes/aes-x64/X64.FastMul.fst(83,3-83,19): (Error 23) user tactic failed: apply_lemma: Cannot instantiate lemma CanonCommSemiring.semiring_reflect CanonCommSemiring.const_last
...`

The hypothesis is that it's failing due to an inability to unify with an if-else expression present.

To test this hypothesis, we tried changing `src/crypto/aes/aes-x64/X64.FastMul.vaf` to wrap the if-else in a function.  You can try this by commenting out line 128 and uncommenting line 129 and rerunning scons.  This version fails with an F* internal error:
`<dummy>(0,0-0,0): (Error 230) Variable "__fname__mem#303637" not found`

@nikswamy thought @mtzguido and @catalin-hritcu might be interested in this issue.  Adding @Chris-Hawblitzel as well, since this came up in the context of our Vale meeting this morning.

Please let me know if you need any help with the repro or with understanding what the code is trying to do.

-Bryan

**Comment by nikswamy** [comment]
Thanks for taking a look and fixing the bug, Guido.

I'm surprised that this code is using pointwise, which is very memory hungry especially in a large local context. Where is the pointwise coming from? I would suggest revising it to use topdown_rewrite, if possible.

Also, it may be possible that the memory explosion is coming from a normalization perf bug that I've been trying to fix on nik_whnf_scrutinee ... But haven't completed.

Nik

Get Outlook for Android<https://aka.ms/ghei36>

________________________________

**Comment by parno** [comment]
@mtzguido Thanks for the bug fix!  I do expect the tactic to finish in a reasonable amount of time, as I've tried in a similar, slightly cleaner context, and it succeeds there.  This could be the normalization bug @nikswamy is working on, or perhaps it's something tactic-specific.

**Comment by mtzguido** [comment]
Actually it was calling `topdown_rewrite`, I was confused since a debug message was saying `Pointwise`. 

As for the bug, now I see this is actually looping, and I think it's somehow building an infinite term (through uvars). When I run it with `--debug_level Norm` it completely halts at one point and starts eating up RAM. Without, it doesn't stop and keeps showing typechecking debug messages, again eating RAM unboundedly. My guess is that in the first case it gets stuck at printing the term.

**Comment by nikswamy** [comment]
Hmm, sounds troublesome. A missing occurs check?

**Comment by mtzguido** [comment]
Sorry, actually I was wrong. It was taking an enormous amount of time while resugaring but I'm not sure it was actually looping. (FWIW, I did some measurements and found that `compress` calls were taking a really long time, multiple seconds, for some `Tm_delayed-resolved` nodes; it might be interesting to see if these can be improved.)

Also the huge terms were indeed coming from pointwise, via `unfold_def`, which has already been troublesome. To get around this I added a new normalizer step called `UnfoldFully` which is similar to `UnfoldOnly` but will continue unfolding the thing recursively.

This seems to work much better. But too much better it seems, since it's causing some failures in `X64.FastMul`:
```
obj/crypto/aes/aes-x64/X64.FastMul.fst(83,3-83,19): (Error 23) user tactic failed: trefl: not a trivial equality (Memory_i_s.buffer_read dst_b 0 (Mkstate?.mem b) +
0x10000000000000000 * Memory_i_s.buffer_read dst_b 1 (Mkstate?.mem b) +
0x100000000000000000000000000000000 *
(Mkstate?.regs b (X64.Machine_s.R11) +
  X64.FastMul.my_bool_to_nat (X64.Vale.Decls_i.overflow (Mkstate?.flags b))) vs Memory_i_s.buffer_read dst_b
  0
  (let { ok = _ ; regs = _ ; xmms = _ ; flags = _ ; mem = __fname__mem } = b in
    __fname__mem) +
0x10000000000000000 *
Memory_i_s.buffer_read dst_b
  1
  (let { ok = _ ; regs = _ ; xmms = _ ; flags = _ ; mem = __fname__mem } = b in
    __fname__mem) +
0x100000000000000000000000000000000 *
((let { ok = _ ; regs = __fname__regs ; xmms = _ ; flags = _ ; mem = _ } = b in
    __fname__regs) (X64.Machine_s.R11) +
  (if
      X64.Vale.Decls_i.overflow (let
            { ok = _ ; regs = _ ; xmms = _ ; flags = __fname__flags ; mem = _ }
          =
            b
          in
          __fname__flags)
    then 1
    else 0))) (see also /home/guido/r/fstar/ulib/FStar.Tactics.Effect.fst(86,32-86,42))
1 error was reported (see above)
```
due to the projectors being unfolded on one side and not the other. It's hard to pinpoint the discrepancy, since the unfolding logic is a bit unclear to me in the normalizer. @nikswamy, there's currently an `unfold_until : option<S.delta_depth>` in `fsteps` and a `delta_level : list<Env.delta_level>` in the `cfg`, and I'm not sure on what each of them represents; should there be a single one, maybe?

Also this should succeed if we properly unified matches, which maybe we can tackle next week?

For now I tried slapping a `compute ()` before the `trefl` calls in `CanonCommSemiring`, and the tactic then succeeds (but a resulting query fails afterwards). So maybe that can get you guys going...

(merging to master in a few minutes)

**Comment by aseemr** [comment]
@mtzguido, i haven't seen the complete code changes yet. but `Unfold_only` also does recursive unfolding, right?

**Comment by mtzguido** [comment]
To clear up the difference:

- `UnfoldOnly` will descend into the definition of the symbol and continue normalizing as usual, possibly unfolding more things.
- `UnfoldFully` will descend into the definition, but changing the configuration to normalize absolutely everything.

So for
```
let x = 2
let y = x + x
```
we have that
```
norm_term [delta_only [%`y]] (`y)   ==  `(x + x)
```
and
```
norm_term [delta_fully [%`y]] (`y)  == `(2 + 2)
```

**Comment by aseemr** [comment]
Thanks for the explanation, that helps!

But can't we do it in the user-land itself, using `topdown_rewrite`? E.g.

```
module Test

open FStar.Tactics

let foo (l:list int) = List.Tot.length l
let bar (l:list int) = List.Tot.length l

//fully unfold the application of term t to n arguments
let unfold_fully (t:term) (n:int) =
  let should_rewrite (s:term) :Tac (bool * int) =
    let hd, tl = collect_app s in
    (term_eq hd t) && List.Tot.length tl = n, 0
  in
  let rewrite () :Tac unit =
    compute ();  //does delta for all
    trefl ()
  in
  topdown_rewrite should_rewrite rewrite

//unfolds foo but not bar
let _ = assert_by_tactic (foo [1;2] == 2 /\ foo [1; 2; 3] == 3 /\ bar [1] == 1) (fun () -> unfold_fully (`foo) 1; dump "A")
```

Or may be I misunderstood? (Just trying to see if we can avoid another normalizer flag.)

**Comment by mtzguido** [comment]
We could, that's in fact similar to the old `unfold_def` (expect that was using `pointwise`).

But the trouble is that `topdown_rewrite` doesn't descend to every subpart of a term (such as lets, matches, and a lot more), and it's also quite expensive. We could (and should!) make it descend to everything, but it will still be pretty expensive due to all the extra goals, typechecking, and unification.

I'd also like to not add a normalizer flag, but it seems reasonable to me here. I expect it to be orders of magnitude faster.

**Comment by aseemr** [comment]
I guess it's fine, I was just curious to understand more.

I buy the incompleteness argument, but a bit fuzzy on the performance one. `topdown_rewrite`, unlike `pointwise`, won't generate redundant unification problems since it uses the callback, which I agree are additional reduction steps. Even for the unification problems `t = ?u` in this particular case of `unfold_fully`, `t` may not have nested uvars(?) and once we are in the `rewrite` above, it's just normalizing.

**Comment by mtzguido** [comment]
I made the changes to make matches unify, and some changes to the delta depths of projectors and their unification, and both alternatives now seem to be working (with an assertion failure at the end, which I'm assuming is unrelated). I noticed the second one is much slower for some reason, by the way.

The bad news is that they cause a bunch of regressions (so this hasn't hit `master`). We just spoke a bit in the developer's meeting about a possible fix. I'll come back with news soon.

**Comment by parno** [comment]
Cool, thanks for the update!  Sounds like you're making progress.

**Comment by mtzguido** [comment]
I think I finally have a (partial..) fix, sorry this took SO long. It required changes in the unifier that took me a lot to get right. I think the current logic makes sense, and it passes existing regressions.

For this file, the tactic runs as expected (with both alternatives) but I still get the failing assertion at the end. I now see this seems to be due to the new unfolding, since the query in question is:
```
obj/crypto/aes/aes-x64/X64.FastMul.fst(32,0-92,16): (Info) Before calling solver VC=
(*Could not prove goal #1*)
Memory_i_s.buffer_read dst_b
  0
  (let { ok = _ ; regs = _ ; xmms = _ ; flags = _ ; mem = __fname__mem } = b in
    __fname__mem) +
0x10000000000000000 *
Memory_i_s.buffer_read dst_b
  1
  (let { ok = _ ; regs = _ ; xmms = _ ; flags = _ ; mem = __fname__mem } = b in
    __fname__mem) +
(0x100000000000000000000000000000000 *
(let { ok = _ ; regs = __fname__regs ; xmms = _ ; flags = _ ; mem = _ } = b in
  __fname__regs) (X64.Machine_s.R11) +
0x100000000000000000000000000000000 *
(if
    X64.Vale.Decls_i.overflow (let
        { ok = _ ; regs = _ ; xmms = _ ; flags = __fname__flags ; mem = _ } =
          b
        in
        __fname__flags)
  then 1
  else 0)) ==
Memory_i_s.buffer_read inA_b
  0
  (let { ok = _ ; regs = _ ; xmms = _ ; flags = _ ; mem = __fname__mem } = va_s0 in
    __fname__mem) *
Memory_i_s.buffer_read inB_b
  0
  (let { ok = _ ; regs = _ ; xmms = _ ; flags = _ ; mem = __fname__mem } = va_s0 in
    __fname__mem) +
0x10000000000000000 *
(Memory_i_s.buffer_read inA_b
  0
  (let { ok = _ ; regs = _ ; xmms = _ ; flags = _ ; mem = __fname__mem } = va_s0 in
    __fname__mem) *
Memory_i_s.buffer_read inB_b
  1
  (let { ok = _ ; regs = _ ; xmms = _ ; flags = _ ; mem = __fname__mem } = va_s0 in
    __fname__mem))
```
which contains a bunch of unfolded projectors (note this query is an unsolved tactic goal).

I guess the SMT chokes when presented with all this, but I'm not really sure. I'll keep looking and reply.

**Comment by mtzguido** [comment]
Pretty sure it's `CanonCommSemiring` that's unfolding these things, but I couldn't find the exact spot...

**Comment by parno** [comment]
Where are your changes checked in?  Perhaps @Chris-Hawblitzel  and I should take a look at the SMT-side of things.

**Comment by mtzguido** [comment]
In `master` now

**Comment by parno** [comment]
How did you arrive at such a compact description of the goal?  I tried changing my local definition of the tactic to be `let int_canon = fun _ -> canon_semiring int_cr; dump "final"`, but it dumps a huge amount of info, starting with:
```proof-state: State dump @ depth 1 (done):
Location: /Users/parno/git/FStar/examples/tactics/CanonCommSemiring.fst(327,12-327,23)
ACTIVE goals (0):

SMT goals (3):
(va_b0: X64.Vale.Decls_i.va_code), (va_s0: X64.Vale.Decls_i.va_state), (dst_b: X64.Machine_s.buffer64), (inA_b: X64.Machine_s.buffer64), (inB_b: X64.Machine_s.buffer64), (p: Prims.pure_post (X64.Vale.State_i.state * X64.Vale.Decls_i.va_fuel)), (_:
Prims.squash (X64.Vale.Decls_i.va_require_total va_b0 (X64.FastMul.va_code_fast_multiply ()) va_s0 /\
    Mkstate?.ok va_s0 /\
    x:
    Prims.unit
...
```
and ultimately it reports `Could not prove goal #1`.  It doesn't seem to be a timeout, in that increasing the rlimit doesn't cause it to take any longer than before.  I also tried enabling Z3's NLarith, but that doesn't seem to help either.  

Looking at your output above, it appears to be the correct, canonical expression, and it more or less lines up with the version as stated in the lemma in `FastMul_helpers_i.fst`, except for all of the let expansions, but the lemma goes through and the in situ version still doesn't.  Not sure if it's a tactic issue at this point, or just Z3 being unhelpful.

**Comment by mtzguido** [comment]
I get that output from `--debug_level SMTQuery`, `--debug_level Rel` or `--debug_level Tac`, any of those will print that bit. Z3 / our encoding should probably prove this one, but hopefully we should be able to control unfolding as well. I'll work a bit more on it tonight and report.

**Comment by parno** [comment]
Wow, this is proving to be quite a doozy.  Thanks for sticking with it!

**Comment by mtzguido** [comment]
Fixed the unification issue, now it's unfolding too much again! But at least I have a better clue:

Somehow, the `vmap` in `CanonCommMonoid` is getting way too unfolded. The bad part is that it's out of the control of tactics since the `vmap` is a pure function from values to values, instead of returning *syntax*. If we could make them have `term` for a codomain, this issue would disappear completely since we would be in full control of normalization, so that's something we could explore.

Nevertheless, it's weird that it's normalizing so much, I think. So I'll look some more and see if something is fishy.

**Comment by mtzguido** [comment]
This actually seems to work in `nik_rel`, which should be landing into `master` soon, so this might just go up and running then! I think there's still a lingering issue, but we might be able to continue for now. I'll keep you posted.

**Comment by parno** [comment]
Interesting that it's working with Nik's fixes.  That's good enough to let us keep making progress on our side at least.

**Comment by mtzguido** [comment]
Everything is now in master, btw.

**Comment by mtzguido** [comment]
For the record, we're now looking at this from the z3 side. The extra unfolding doesn't really seem to make a difference, and it's really hard to remove it given how `CanonCommSemiring` works, maybe plain  impossible.

The resulting query seems to involve some NL arith, which might be stumping z3 given the big context (even if pruning with the `prune` tactic). According to Nikolaj Bjørner, z3 might get improved support for (real) NL arith which we could leverage, but that's down the line...

**Comment by parno** [comment]
Do you think we can massage the context further (possibly with something Vale-specific) to arrive at something isomorphic to the helper lemma context?  Or is it time to give up and decide that F*/Z3 can't handle this kind of query at this point, and so we should fallback to manually writing out the full version of the helper lemma?

**Comment by mtzguido** [comment]
This turned out to be a weird encoding issue, and was fixed by 63fabb782639c0a93de8c40eacf1ec74a8f8413b. We just forgot to close this.

**Comment by parno** [comment]
Cool, thanks for the update.
